lms/services/assignment.py

```diff
@@ -32,6 +34,16 @@ def get_assignment(self, tool_consumer_instance_guid, resource_link_id):
             .one_or_none()
         )
 
+    def get_assignments_for_grouping(self, grouping_id) -> List[Assignment]:
```

Comment:
I reckon we might have to revisit this at some point. I still think we'll get  more flexible and easiert to compose  services if we remove the `.all` here and return a query instead.

The joins here will be about what data you get, any eager-loading applied outside would be about how you get it.

That's in any case a bigger discussion for another moment.

Reply:
I thought having about having a lazy load spec object with a bunch of booleans on it so you could say what you wanted but:

* The way SQLAlchemy makes you do it, it's a real pain to layout and would be a nest of ifs
* I think we'd only ever call it in one configuration, so all that conditionality would have to be tested, but would only be in one configuration in actual usage

It's a tough one. I can see why giving control to the caller would be good from a full control point of view, but I'm not sure you have much of a service boundary in that world. The application of the correct eager loading etc. is quite difficult to get right (harder than the query here). I'm not sure what kind of encapsulation the service would be providing then? It becomes pretty close to just letting everyone do their own ORM queries.

One possible more radical concept is we might have a missing layer? I kind of hate to suggest this, but pre-baked queries as their own thing, without specifying how they are used? This isn't a fleshed out idea, but maybe there's something between services and models? Dunno.

In the next few PR's we'll be starting to bring eager loading in. My plan to start with was you'd ask `eager=True` and have the full shebang applied. I'm not sure we'll ever have a second caller for most of these things? So maybe if we ever do, and that second caller want's a different set of stuff we tackle it then?

Reply:
Yeah I don't have a concrete proposal and I agree  with this point 

> I'm not sure what kind of encapsulation the service would be providing then

But I think we are already in a world were that encapsulation is already very thin, we return ORM objects from the service and the views/other services are free to have DB side effects with them.

SQLA is huge and I think trying to build a layer around it would be mistake. Is the alternative to have all the views using it freely? Probably not.

My preferred solution in other projects has been to have other plain python objects (data-classes?) as the return values when the service knows how many objects it's returning (typically, one) and return queries when the number is unknown and >1. 

This brings a lot of duplication with those objects that are very similar to the models but helps avoiding a lot of problems with tracking where the queries happens and composes very well in the query side, for example implementing pagination can be done without shoving all that logic into the services.